src/crystal/system/unix/file_descriptor.cr

```diff
@@ -0,0 +1,145 @@
+require "c/fcntl"
+
+# :nodoc:
+module Crystal::System::FileDescriptor
+  include IO::Syscall
+
+  @fd : Int32
+
+  @read_event : Event::Event?
+  @write_event : Event::Event?
+
+  private def unbuffered_read(slice : Bytes)
+    read_syscall_helper(slice, "Error reading file") do
+      # `to_i32` is acceptable because `Slice#size` is a Int32
+      LibC.read(@fd, slice, slice.size).to_i32
+    end
+  end
+
+  private def unbuffered_write(slice : Bytes)
+    write_syscall_helper(slice, "Error writing file") do |slice|
+      LibC.write(@fd, slice, slice.size).tap do |return_code|
+        if return_code == -1 && Errno.value == Errno::EBADF
+          raise IO::Error.new "File not open for writing"
+        end
+      end
+    end
+  end
+
+  private def system_blocking?
+    fcntl(LibC::F_GETFL) & LibC::O_NONBLOCK == 0
+  end
+
+  private def system_blocking=(value)
+    current_flags = fcntl(LibC::F_GETFL)
+    new_flags = current_flags
+    if value
+      new_flags &= ~LibC::O_NONBLOCK
+    else
+      new_flags |= LibC::O_NONBLOCK
+    end
+    fcntl(LibC::F_SETFL, new_flags) unless new_flags == current_flags
+  end
+
+  private def system_close_on_exec?
+    flags = fcntl(LibC::F_GETFD)
+    (flags & LibC::FD_CLOEXEC) == LibC::FD_CLOEXEC
+  end
+
+  private def system_close_on_exec=(arg : Bool)
+    fcntl(LibC::F_SETFD, arg ? LibC::FD_CLOEXEC : 0)
+    arg
+  end
+
+  def self.fcntl(fd, cmd, arg = 0)
+    r = LibC.fcntl(fd, cmd, arg)
+    raise Errno.new("fcntl() failed") if r == -1
+    r
+  end
+
+  def fcntl(cmd, arg = 0)
+    FileDescriptor.fcntl(@fd, cmd, arg)
+  end
+
+  private def system_stat
+    if LibC.fstat(@fd, out stat) != 0
+      raise Errno.new("Unable to get stat")
+    end
+    ::File::Stat.new(stat)
+  end
+
+  private def system_seek(offset, whence : IO::Seek) : Nil
+    seek_value = LibC.lseek(@fd, offset, whence)
+
+    if seek_value == -1
+      raise Errno.new "Unable to seek"
+    end
+  end
+
+  private def system_pos
```

Comment:
I'd prefer `system_tell` over `system_pos`.

I wish we'd drop `IO#pos` and `IO#pos=` and kept `IO#seek` and `IO#tell` instead (less duplication, more explicit naming).

Reply:
We just need 1 call: `seek` which always returns the current position. I suggested that `IO` implement `pos` `pos=` and `tell` in the base class to @asterite and simply provide a raising `IO#tell` by default.

For now, this is how `IO` works. It's not great, but it is how it is until I refactor.

Reply:
Personally i'd rather keep `pos`, `pos=` and `seek`. `tell` is obscure and weird naming to me.

Reply:
Well, it's "seek and tell". They're well known associated terms, whatever the language. E.g. C, Haskell, Python, Ruby or Rust, all have both.

I don't like the `pos` naming, it's an abbreviation and inexplicit (inherited from Ruby). At least `position` would be a more explicit name but I still prefer `tell` (broadly used). Using a setter (`pos=`) to perform an action,  that is move the cursor in a file, sounds awful to me, in addition to be an exact duplicate for `seek`, too.

Reply:
I have to say "seek" is not an unknown term to me, although "tell" is. I agree with you on `pos` vs `position`, but not that setters shouldn't have performance side-effects.

Reply:
As far as I can tell, the terms tell and seek come from working with files. `IO` is a general class, and to me it doesm't seem logical to transfer this meaning from files to any other stream. But I may be mistaken.
Another thought: `pos` and `pos=` is a concept easy to understand. One reads the current stream position, the other sets it (`position` and `position=` are a bit longer but also fine). With `seek` and `tell` it is more complicated, if your not familiar with these terms. You have to remember two names instead of one and can't simply asseses their meaning.

Reply:
I'm also for dropping `#seek/#tell` and using `#position/#position=` (Not `#pos`).  The latter is really clear in intention.  Also, it lets you seek around in a file using normal code:

```cr
my_file.position += 5 # This is how you'd write it anywhere else too! Fantastic!
my_file.seek(my_file.tell + 5) # Compared to the one above: Meh.
my_file.seek(5, File::SEEK_CUR) # I can never remember which SEEK_* I need - this sucks.
```

Reply:
I think `seek` should stay, `io.position = io.size - 5` involves 2 syscalls whereas `io.seek(-5, IO::Seek::End)` requires only one. Similar for `io.position += 5`. That being said, I think that using `io.position` is a much nicer interface which is more than adequate for most people. Perhaps some caching for size and position would mean that we can obviate `seek` entirely.

Regardless of that, we should make a new issue containing all the IO refactor suggestions from this thread. Unless anyone else wants to do it now, i'll get it done once this PR is merged.